### ISSUE TITLE
Consider making recursion optional

### ISSUE THREAD

**Issue by stepancheg** [bazelbuild/starlark#97]
The spec prohibits recursion:

> Execution is finite
> The language does not allow recursion

However:

* recursion does not break Starlark model of deterministic hermetic evaluation
* prohibited recursion only helps to achieve finite execution in theory, while in practice it's easy to write 5 lines program which won't finish before the end of the Universe
* there is nothing in Starlark language or implementations to my knowledge which depends on the fact that the execution is finite

We find recursion useful, and we allow it in our uses of Starlark. We would like to stay within the Starlark spec.

Is it possible to make recursion optional in Starlark language? And Bazel Starlark implementation could still restrict it.

**Comment by alandonovan** [comment]
I think it's completely reasonable to make this an option, as the implementation at go.starlark.net does.

**Comment by alandonovan** [comment]
I agree that prohibiting recursion only stops people in theory. Many theoretically expensive computations run quickly, and many bounded computations are slow. Committed users can (and do) implement a Turing machine with a large but bounded number of steps. What really matters is how many steps a computation takes, and that is best measured or limited dynamically. (We already have measurement in place, and we plan to add limits.)

That said, prohibiting recursion has worked fairly well as a signal to users that they are attempting the kinds of programming perhaps best not done in the build system.

Care to send a PR to the Java implementation? The necessary change would add a boolean to StarlarkSemantics and skip the recursion check if the flag is set. The Go implementation goes a step further and adds a `while` loop to the grammar. To do that in the Java implementation would require adding a new statement type, changing the parser, and adding a FileOption to enable it.

@laurentlb Any concerns here?

**Comment by stepancheg** [comment]
I'll send PR against Bazel repo to allow recursion with option.

@ndmitchell do we want `while` loops? I'm fine either way, but better all implementations to implement the same language.

**Comment by ndmitchell** [comment]
Why is there a requirement to patch the Java implementation and the spec simultaneously? Should the spec live as its own artefact, and then implementation changes to Java to conform to the spec (or for Java to choose not to adhere closely) are a separate thing? (Genuine question - still trying to figure out where this spec lies for everyone else.)

@stepancheg - no objection to while loops. If you have recursion anyway, it's often cleaner than expressing the same thing without a while loop. It seems like a natural inclusion.

**Comment by alandonovan** [comment]
There's no such requirement; in fact, the spec lives in a GitHub repo whereas the Java code lives in Google's internal repo and is mirrored to GitHub, so they can't change atomically. Arguably it would be cleaner if there were two documents, the spec and the description of the Java implementation (like we have for Go), the first being the statement of intent and the second being the ground truth, but maintaining two docs in two repos is already a lot of work. In the absence of that hypothetical doc, the spec serves as the reference for Bazel users, so it's better if our aspirations don't run too far ahead of our implementations.

**Comment by alandonovan** [comment]
Re: while loops: let's make it a separate PR.

**Comment by ndmitchell** [comment]
Would it make sense to have the spec detail in-place where the various implementations diverge from the spec, and in what ways? Reducing it to one doc in one repo might halve your effort :)

**Comment by alandonovan** [comment]
I don't think the spec should describe the particulars of any one implementation (there are three).

**Comment by ndmitchell** [comment]
I don't think the spec should describe the particulars, but I've found things like the MDN approach e.g. at https://developer.mozilla.org/en-US/docs/Web/CSS/:default of describing the spec, and then briefly saying what is conformant/non-conformant to be very helpful. Certainly not essential though if you'd rather not include it.

**Comment by alandonovan** [comment]
I suspect the CSS team at Microsoft is rather bigger than the Starlark subteam of the Bazel team at Google.

**Comment by stepancheg** [comment]
Spec PR https://github.com/bazelbuild/starlark/pull/99

**Comment by laurentlb** [comment]
Some notes about recursion: yes, it's fine to allow it as an optional extension. But I'd rather not enable it in Bazel.


Starlark should remain usable as a configuration language. It's common for a config language to prevent infinite loops, e.g. Dhall (http://www.haskellforall.com/2020/01/why-dhall-advertises-absence-of-turing.html). As Alan wrote, it's used as a signal to discourage users to write things that don't belong to a build system / configuration language.

Recursion can also make the code more difficult to understand (both for human readers and static analysis tools) in a Python-like language. I tried to make Starlark minimalist, reduce the number of concepts users to understand, and force everyone to use the same patterns (this can be important when you maintain a large codebase). I believe the lack of recursion fits with those goals.

When we migrated Google codebase from Python macros to Starlark, we had to convert many Python files. The lack of recursion was not a concern for the migration. Since that, a large amount of Starlark code was written; very few people asked about recursion. Based on these data points, recursion doesn't seem like an important feature to us.

**Comment by konste** [comment]
@laurentlb: Recursion can also make the code more difficult to understand

In order to understand what is going on between custom rules I had to implement something akin pretty_print for Starlark to be able to make sense of the random objects. Implementation with recursion would be transparent and easy to understand. Without recursion I had to implement  "poor man recursion" using list as a stack and believe me it is MUCH more difficult to understand than if it would be done with the real recursion. 
Example fragment: 
``` 
    state = struct(obj = obj, name = name, prefix = "") 
    stack = [state] # list has methods append and pop
    for _ in range (2147483647):
        state = stack.pop() 
```
Isn't it ugly? 
Each feature may be used and may be abused. In case of recursion I'd consider it a must for any modern language.
